src/spark/state.cpp

```diff
@@ -0,0 +1,1333 @@
+    spark::Coin txCoin(coin.params);
+    // cycle transaction hashes, looking for this coin
+    for (CTransactionRef tx : block.vtx){
+        uint32_t nIndex = 0;
+        for (const CTxOut &txout : tx->vout) {
+            if (txout.scriptPubKey.IsSparkMint() || txout.scriptPubKey.IsSparkSMint()) {
+                try {
+                    ParseSparkMintCoin(txout.scriptPubKey, txCoin);
+                }
+                catch (...) {
+                    continue;
+                }
+                if(coin == txCoin){
+                    outPoint = COutPoint(tx->GetHash(), nIndex);
+                    return true;
+                }
+            }
+            nIndex++;
+        }
+    }
+    return false;
+}
+
+static bool CheckSparkSpendTAg(
+        CValidationState& state,
+        CSparkTxInfo* sparkTxInfo,
+        const GroupElement& tag,
+        int nHeight,
+        bool fConnectTip) {
+    // check for spark transaction in this block as well
+    if (sparkTxInfo &&
+        !sparkTxInfo->fInfoIsComplete &&
+        sparkTxInfo->spentLTags.find(tag) != sparkTxInfo->spentLTags.end())
+        return state.DoS(0, error("CTransaction::CheckTransaction() : two or more spark spends with same tag in the same block"));
+
+    // check for used tags in sparkState
+    if (sparkState.IsUsedLTag(tag)) {
+        // Proceed with checks ONLY if we're accepting tx into the memory pool or connecting block to the existing blockchain
+        if (nHeight == INT_MAX || fConnectTip) {
+            return state.DoS(0, error("CTransaction::CheckTransaction() : The Spark spend tag has been used"));
+        }
+    }
+    return true;
+}
+
+/******************************************************************************/
+// CLelantusState
+/******************************************************************************/
+
+CSparkState::CSparkState(
+        size_t maxCoinInGroup,
+        size_t startGroupSize)
+        :
+        maxCoinInGroup(maxCoinInGroup),
+        startGroupSize(startGroupSize)
+{
+    Reset();
+}
+
+void CSparkState::Reset() {
+    coinGroups.clear();
+    latestCoinId = 0;
+    mintedCoins.clear();
+    usedLTags.clear();
+    mintMetaInfo.clear();
+    spendMetaInfo.clear();
+}
+
+std::pair<int, int> CSparkState::GetMintedCoinHeightAndId(const spark::Coin& coin) {
+    auto coinIt = mintedCoins.find(coin);
+
+    if (coinIt != mintedCoins.end()) {
+        return std::make_pair(coinIt->second.nHeight, coinIt->second.coinGroupId);
+    }
+    return std::make_pair(-1, -1);
+}
+
+bool CSparkState::HasCoin(const spark::Coin& coin) {
+    return mintedCoins.find(coin) != mintedCoins.end();
+
+}
+
+bool CSparkState::HasCoinHash(spark::Coin& coin, const uint256& coinHash) {
+    for (auto it = mintedCoins.begin(); it != mintedCoins.end(); ++it ){
+        const spark::Coin& coin_ = (*it).first;
+        if(primitives::GetSparkCoinHash(coin_) == coinHash){
```

Comment:
long standing problem with code style: `if` should be followed by space everywhere. Nitpicking but it'll be prettier if it was the same

Reply:
Wow, it's been a while since I've seen an if statement without a space 🤣 It about gave me a stroke

Reply:
fixed